opm/common/quad.hpp

```diff
@@ -0,0 +1,325 @@
+template <>
+struct is_standard_layout<quad>
+    : public integral_constant<bool, true>
+{};
+
+template <>
+struct is_trivial<quad>
+    : public integral_constant<bool, true>
+{};
+
+/*
+template <>
+struct is_trivially_copyable<quad>
+    : public integral_constant<bool, true>
+{};
+*/
+
+template <class OtherType>
+struct is_assignable<quad, OtherType>
+    : public integral_constant<bool, is_arithmetic<OtherType>::value>
+{};
+
+template <class OtherType>
+struct is_nothrow_assignable<quad, OtherType>
+    : public is_assignable<quad, OtherType>
+{};
+
+/*
+template <class OtherType>
+struct is_trivially_assignable<quad, OtherType>
+    : public integral_constant<bool, is_arithmetic<OtherType>::value>
+{};
+*/
+
+template <>
+struct is_copy_assignable<quad>
+    : public integral_constant<bool, true>
+{};
+
+template <>
+struct is_nothrow_copy_assignable<quad>
+    : public integral_constant<bool, true>
+{};
+
+template <>
+struct is_move_assignable<quad>
+    : public integral_constant<bool, true>
+{};
+
+template <>
+struct is_nothrow_move_assignable<quad>
+    : public integral_constant<bool, true>
+{};
+
+template <>
+struct is_constructible<quad>
+    : public integral_constant<bool, true>
+{};
+
+template <>
+struct is_nothrow_constructible<quad>
+    : public integral_constant<bool, true>
+{};
+
+template <>
+struct is_default_constructible<quad>
+    : public integral_constant<bool, true>
+{};
+
+template <>
+struct is_nothrow_default_constructible<quad>
+    : public integral_constant<bool, true>
+{};
+
+/*
+template <>
+struct is_trivially_default_constructible<quad>
+    : public integral_constant<bool, true>
+{};
+*/
+
+template <>
+struct is_copy_constructible<quad>
+    : public integral_constant<bool, true>
+{};
+
+template <>
+struct is_move_constructible<quad>
+    : public integral_constant<bool, true>
+{};
+
+template <>
+struct is_nothrow_move_constructible<quad>
+    : public integral_constant<bool, true>
+{};
+
+
+template <>
+struct is_destructible<quad>
+    : public integral_constant<bool, true>
+{};
+
+template <>
+struct is_nothrow_destructible<quad>
+    : public integral_constant<bool, true>
+{};
+
+template <class OtherType>
+struct is_convertible<quad, OtherType>
+    : public is_arithmetic<OtherType>
+{ };
+
+inline std::ostream& operator<<(std::ostream& os, const quad& val)
+{ return (os << double(val)); }
+
+inline std::istream& operator>>(std::istream& is, quad& val)
+{
+    double tmp;
+    std::istream& ret = (is >> tmp);
+    val = tmp;
+    return ret;
+}
+
+inline quad abs(quad val)
+{ return (val < 0) ? -val : val; }
+
+inline quad floor(quad val)
+{ return floorq(val); }
+
+inline quad ceil(quad val)
+{ return ceilq(val); }
+
+inline quad max(quad a, quad b)
+{ return (a > b) ? a : b; }
+
+inline quad min(quad a, quad b)
+{ return (a < b) ? a : b; }
+
+inline quad sqrt(quad val)
+{ return sqrtq(val); }
+
+template <class ExpType>
+inline quad pow(quad base, ExpType exp)
+{ return powq(base, exp); }
```

Comment:
To see if I got this right: pow(quad, double) is ambiguous because the compiler cannot choose between pow(double, double) and pow(quad, quad)? I'm a little surprised, I would have thought the conversion direction with no loss of data would be preferred.

In any case, go ahead with the template, but can you test first that it won't match pow(double, double) calls?